Source/dafny.msbuild/DafnyVersionCheckTask.cs

```diff
@@ -0,0 +1,157 @@
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using Microsoft.Build.Framework;
+using Task = Microsoft.Build.Utilities.Task;
+
+namespace DafnyMSBuild
+{
+    /**
+     * A custom Task for checking that the provided Dafny CLI is of a compatible version.
+     */
+    public class DafnyVersionCheckTask : Task
+    {
+        public string DafnyBinariesDir { get; set; }
+
+        [Required]
+        public string DafnyExecutablePath { get; set; }
+
+        // Either a Dafny version number, like "2.3.0.10506", or the empty string.
+        public string DafnyVersion { get; set; }
+
+        // Either a Git commit hash, or the empty string.
+        public string DafnyVersionCommit { get; set; }
+
+        // Either a Git tag name, or the empty string.
+        public string DafnyVersionTag { get; set; }
+
+        public override bool Execute()
+        {
+            var foundDafnyVersion = GetDafnyExecutableVersion(DafnyExecutablePath);
+
+            var enforceVersion = !string.IsNullOrEmpty(DafnyVersion);
+            var enforceCommit = !string.IsNullOrEmpty(DafnyVersionCommit);
+            var enforceTag = !string.IsNullOrEmpty(DafnyVersionTag);
+            if (!(enforceVersion || enforceCommit || enforceTag)) return Success(foundDafnyVersion);
+            if (new[] {enforceVersion, enforceCommit, enforceTag}.Count(b => b) > 1) {
+                return Fail("The DafnyVersion, DafnyVersionCommit, and DafnyVersionTag properties are mutually exclusive");
+            }
+
+            if (enforceVersion) return DafnyVersion == foundDafnyVersion
+                ? Success(foundDafnyVersion)
+                : Fail($"Expected Dafny version \"{DafnyVersion}\", found \"{foundDafnyVersion}\"");
+
+            var expectedRepoHead = enforceCommit ? $"commit {DafnyVersionCommit}" : $"tag {DafnyVersionTag}";
+            if (string.IsNullOrEmpty(DafnyBinariesDir))
+            {
+                return Fail("DafnyBinariesDir must be set to a directory within a Git repo in order to check DafnyVersionCommit or DafnyVersionTag");
+            }
+            if (IsGitWorkingTreeDirty(DafnyBinariesDir))
+            {
+                return Fail($"Expected Dafny executable Git repo at {expectedRepoHead}, but working tree or index has changes");
+            }
+
+            var specifiedCommitHash = enforceCommit ? DafnyVersionCommit : GetCommitHashForGitTag(DafnyBinariesDir, DafnyVersionTag);
+            var repoHeadCommitHash = GetGitRepoHeadCommitHash(DafnyBinariesDir);
+            return repoHeadCommitHash != null && repoHeadCommitHash == specifiedCommitHash
+                ? Success(foundDafnyVersion)
+                : Fail($"Expected Dafny executable Git repo at {expectedRepoHead}, but HEAD is at {repoHeadCommitHash}");
+        }
+
+        /**
+         * Prints a success message and returns true.
+         */
+        private bool Success(string version)
```

Comment:
I would make the messages more specific to whichever of the three options are active here. As is you're logging "Found Dafny version 3.5.0" even if the user specified `<DafnyVersionCommit>abcdef1234...`. I think it would be more reassuring in that case to log "Found Dafny repo at abcdef1234..."

Reply:
Good idea, added in latest commits.